pykern/pkunit.py

```diff
@@ -510,20 +514,44 @@ def _actual_xlsx_to_csv(self, actual_xlsx, sheet):
             line_terminator="\r\n",
         )
 
+    def _gen_ignore_lines(self):
+        r = []
+        if self._ignore_lines is None:
+            return r
+        for l in (l for l in [l for l in self._ignore_lines if l]):
+            r.extend(["-I", l])
+        return r
+
     def _compare(self):
         if self._expect == self._actual:
             return
-        c = f"{'cmp' if self.is_bytes else 'diff'}  '{self._expect_path}' '{self._actual_path}' 2>&1"
-        with os.popen(c) as f:
+
+        c = ["cmp"] if self.is_bytes else ["diff", *self._gen_ignore_lines()]
+        c.extend(
+            [
+                f"{self._expect_path}",
+                f"{self._actual_path}",
+            ]
+        )
+        p = subprocess.run(
+            c,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            text=True,
+        )
+        if p.returncode != 0:
             pkfail(
-                "{}",
                 f"""expect != actual:
-    {c}
-    {f.read()}
-    {self._message()}
-    """,
+'{"' '".join(c)}'
+{p.stdout}
+{self._comparison_message(p.returncode)}""",
```

Comment:
This should be a nested function, too. Describe what it is, and again cohesive modularization. There's an indentation issue in _message (those lines should be left justfied), and combine in a single nested function:

```py
def _failed_msg(process):
        r = "'" + "' '".join(process.args) + f"'\n" + p.stdout + "\n"
        if not (return_code == 1 or self.is_bytes):
            return r + "diff command failed\n"
        if self._expect_is_jinja:
            return r + f"""
Implementation restriction: expect is a jinja template which has been processed to
produce the diff. A simple copy of actual to expect is not possible. You will need to update
the expect jinja template={self._expect_path} manually.
"""
        return r + f"""
to update test data:
    cp '{self._actual_path}' '{self._expect_path}'
"""
```